libraries/lts/include/mcrl2/lts/detail/liblts_bisim_m.h

```diff
@@ -0,0 +1,579 @@
+        }
+    };
+
+    struct formula 
+    {
+        formula_index_type index;
+        label_type label;
+        bool negated;
+        std::vector <formula> conjunctions;
+        std::set <block_index_type > truths;
+
+        int depth() {
+            int max_depth = 0;
+            for (formula f : conjunctions) {
+                if (f.depth() > max_depth) {
+                    max_depth = f.depth() + 1;
+                }
+            }
+            return max_depth; 
+        }
+    };
+
+    std::vector < block > blocks;
+    // Blocks that are split become inactive.
+    std::vector < state_type > block_index_of_a_state;
+    std::vector < bool > block_flags;
+    std::vector < bool > state_flags;
+
+    std::vector< block_index_type > to_be_processed;
+    std::vector< block_index_type > BL;
+    typedef std::pair<label_type, block_index_type> observation_t;
+    typedef std::set <observation_t> derivatives_t;
+    std::map < std::pair<block_index_type, block_index_type>, level_type> greatest_common_ancestor;
+
+
+    /* Post processes the partition structure to save outgoing transitions per block */
+    void save_transitions() {
+        for (transition t : aut.get_transitions()) {
+            
+            block_index_type sourceBlock = block_index_of_a_state[t.from()];
+            block_index_type targetBlock = block_index_of_a_state[t.to()];
+    
+            blocks[sourceBlock].outgoing_observations.insert(std::make_pair(t.label(), targetBlock));
+        }
+    }
+
+    void set_truths(formula& f) {
+        std::set <block_index_type> image_truths;
+        std::set <block_index_type> pre_image_truths;
+        std::set<block_index_type> intersection;
+
+        
+        image_truths = std::set(partition);
+
+        for (formula df : f.conjunctions) {
+            std::set_intersection(image_truths.begin(), image_truths.end(), df.truths.begin(), df.truths.end(),
+                std::inserter(intersection, intersection.begin())
+            );
+            image_truths.swap(intersection);
+            intersection.clear();
+        }
+
+        //Now compute preimage according to label
+        for (block_index_type B : image_truths) {
+            for (transition t : blocks[B].transitions) {
+                if (t.label() == f.label && (pre_image_truths.find(block_index_of_a_state[t.from()]) == pre_image_truths.end()) ) {
+                    pre_image_truths.insert(block_index_of_a_state[t.from()]);
+                }
+            }
+        }
+
+        if (f.negated) {
+            image_truths.swap(pre_image_truths);
+            pre_image_truths.clear();
+            std::set_difference(partition.begin(), partition.end(), image_truths.begin(), image_truths.end(),
+                std::inserter(pre_image_truths, pre_image_truths.begin()));
+        }
+        f.truths.swap(pre_image_truths);
+    }
+
+    void create_initial_partition() {
+        to_be_processed.clear();
```

Comment:
This feels like it is part of the construction as I don't see it being called anywhere else. In that case these assignments could be also be placed where the members are declared above. The rest of the initialisation could then be moved into the constructor.